tests/parsenum/main.c

```diff
@@ -0,0 +1,55 @@
+#include <stdint.h>
+#include <stdio.h>
+
+#include "parsenum.h"
+#include "warnp.h"
+
+int
+main(int argc, char * argv[])
+{
+	double d = 0.0;
+	size_t s = 0;
+	int i = 0;
+
+	(void)argc; /* UNUSED */
+	(void)argv; /* UNUSED */
+
+	WARNP_INIT;
+
+#define TEST(x, y, z, w) do {						\
+	fprintf(stderr, "PARSENUM(\"%s\", %s, %s)\n", y, #z, #w);	\
+	if (PARSENUM(x, y, z, w)) {					\
+		if (errno == EINVAL) {					\
+			warn0("PARSENUM: EINVAL");			\
+		} else if (errno == ERANGE) {				\
+			warn0("PARSENUM: ERANGE");			\
+		} else {						\
+			warnp("PARSENUM");				\
+		}							\
+	} else {							\
+		fprintf(stderr, "%f %zu %d\n", d, s, i);		\
+	}								\
+} while (0)
+
+	TEST(&d, "123.456", 0, 1000);
+	TEST(&s, "1234", 0, 4000);
+	TEST(&i, "12345", -10, 100);
+
+#define TEST2(x, y) do {				\
+	fprintf(stderr, "PARSENUM(\"%s\")\n", y);	\
+	if (PARSENUM(x, y)) {				\
+		if (errno == EINVAL) {			\
+			warn0("PARSENUM: EINVAL");	\
+		} else if (errno == ERANGE) {		\
+			warn0("PARSENUM: ERANGE");	\
+		} else {				\
+			warnp("PARSENUM");		\
+		}					\
+	} else {					\
+		fprintf(stderr, "%f %zu %d\n", d, s, i);\
+	}						\
+} while (0)
```

Comment:
The more I think about it, the more I think that these are not the right way to write testing macros for PARSENUM.  Something more like

    TEST2_SUCCESS("123.456", double, 123.456);
    TEST2_FAILURE("123.456", size_t, EINVAL);
    TEST4_FAILURE("1234", int, -10, 10 ERANGE);

producing output of

    Parsing "123.456" as type double yields 123.456... PASSED!
    Parsing "123.456" as type size_t yields EINVAL... PASSED!
    Parsing "1234" as type int between -10 and 10 yields ERANGE... PASSED!

might be better?